### DISCUSSION TITLE
selecting a newly created object

### DISCUSSION THREAD

**Discussion by alexted** [sqlalchemy/sqlalchemy#11775] [Usage Questions] [answered]
I thought flush would help me, but it didn't work. Commit closes the session - I won't be able to select after insert. 
What can I do in this situation? 
Is it possible to enrich data the object that insert returns to me in the current transaction?

```python
class User(Base):
    __tablename__ = "user"

    id: Mapped[t_UUID] = mapped_column(UUID, primary_key=True, server_default=func.gen_random_uuid())
    first_name: Mapped[str] = mapped_column(String(64), nullable=True)
    last_name: Mapped[str] = mapped_column(String(64), nullable=True)
    email: Mapped[str] = mapped_column(String(320), unique=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=utcnow())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=utcnow(), onupdate=utcnow())
    is_deleted: Mapped[bool] = mapped_column(Boolean, server_default=sql.false())

    organisation_role: Mapped["UserOrganisationRole"] = relationship(
        back_populates="member", lazy="joined", innerjoin=True
    )
    project_roles: Mapped[list["UserProjectRole"]] = relationship(
        back_populates="member", lazy="joined", innerjoin=True
    )

    def __repr__(self) -> str:
        return f"User(id={self.id!r}, email={self.email!r})"
```

```python
    async def create_user(self, data: CreateUserDTO) -> UserDTO:
        stmt = insert(User).values(**data.model_dump()).returning(User)

        user = await self.session.scalar(stmt)

        ###### I can't use session.commit() here.

        stmt = (
            select(User)
            .where(User.id == user.id)
            .join(UserOrganisationRole)
            .options(
                joinedload(User.organisation_role, innerjoin=True).load_only(UserOrganisationRole.role),
                joinedload(User.project_roles, innerjoin=True).load_only(
                    UserProjectRole.project_id, UserProjectRole.role
                ),
            )
        )

        result = await self.session.scalar(stmt) ###### I get None here.

        return self.convert_to_dto(result)
```

**Comment by CaselIT** [comment]
Hi,

You are using `.join(UserOrganisationRole)`. That's an inner join.
Are you sure that `UserOrganisationRole` has data for your newly created user? From your snippet the answer seems no

**Reply by alexted** [reply]
Yeah, they're not really there, you're right. But the thing is, they're not loading at all. An error is thrown when trying to access these attributes.
![image](https://github.com/user-attachments/assets/67d4d29f-c37c-43c0-b610-f870556d7da7)

And the result is None here:
![image](https://github.com/user-attachments/assets/553255f9-eed2-45aa-a1d2-d541b3000a8a)

**Reply by CaselIT** [reply]
I don't follow. If you have no matching UserOrganisationRole, none is the correct value for that query

**Reply by alexted** [reply]
First, I do the insert:
```python
stmt = insert(User).values(**data.model_dump()).returning(User)

user = await self.session.scalar(stmt)
```
The user object I got here causes an error when trying to refer to organization_role and project_roles - they are not None - see the first screenshot from the IDE above.

**Reply by alexted** [reply]
So I do the next one query in the database:
```python
stmt = (
            select(User)
            .where(User.id == user.id)
            .join(UserOrganisationRole)
            .options(
                joinedload(User.organisation_role, innerjoin=True).load_only(UserOrganisationRole.role),
                joinedload(User.project_roles, innerjoin=True).load_only(
                    UserProjectRole.project_id, UserProjectRole.role
                ),
            )
        )

        result = await self.session.scalar(stmt) 
### The result is always None here^^
```

**Reply by alexted** [reply]
I don't do session.commit manually between these queries - this is the problem. The commit closes the session and I can't execute the select. But I need to load more user data. How can I do it?

**Reply by CaselIT** [reply, marked as answer]
Running `session.commit` would not change what the query returns, since that query will return 0 rows, as mentioned above.

If you just want to populate the relationship you could use `insert.options()`, but I'm not 100% sure if joinedload works there. selectinload should work though.

If you want to use an explicit second query then remove the `.join(UserOrganisationRole)` otherwise that query will never return any row. Also remove `innerjoin=True` from the joinedload, otherwise again you get 0 rows back.

You may also need to supply populate_existing=True, otherwise the session will re-use the same user instance that's already loaded. https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#populate-existing

**Reply by alexted** [reply]
The second query returns no such user in principle. It says that there is no such user. I think this is because the user is not physically stored in the database yet. That's why I want to get the necessary data set from the insert at once (and not make selects in principle), or somehow make the second query work and give the necessary data set instead of the truncated one that the insert returns now.

**Reply by CaselIT** [reply]
> I think this is because the user is not physically stored in the database yet

No. As mentioned above that's not because the user is not in the db, it is because you are doing an inner join to a table that as no matching row. Please look at how inner join work https://en.wikipedia.org/wiki/Join_(SQL)#Inner_join

**Comment by alexted** [comment]
Thanks for your help! And one more question, if I may - how to make `organisation_role` come in the form of
`organisation_role: "owner"`
instead of 
`organisation_role: { "role": "owner" }`
I hope I've made it clear what I'm trying to achieve 😅
is there any hack for this?

**Reply by CaselIT** [reply]
You may look into association proxy that will let you map a property of a relationship on the parent class. https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html

**Reply by alexted** [reply]
Thank you!
